netdata/netdata
# feat(cgroups.plugin): add filtering by cgroups names and improve renaming in k8s

##### Summary

So many changes in one PR because the goal was to fix the following problem:

- Netdata loses some containers on a K8s cluster when:
  - the CPU % of the node is high (100% in our case).
  - A lot of containers are created in a short time. 

The changes address that problem. After this patch, the amount of CPU time needed for cgroup name resolution is ~40% less.

`enable by default cgroups names matching` is just a nice addition that came naturally 😄 

---

This PR:

- **[feat]**: Adds `enable by default cgroups names matching` filter. It matches the name of the cgroup (as you see it on the dashboard). It is much easier to filter specific containers, e.g.: `!nginx2 *`.
- **[feat]**: K8s: wait for container initialization (`runc:[2:INIT]` may take a while) and don't try to rename/collect uninitialized containers.
- **[feat]**: K8s: identify [pause containers](https://docs.mirantis.com/mke/3.4/ref-arch/pause-containers.html) and don't try to rename/collect their data. Every pod has a pause container, so it is up to 50% fewer cgroup-rename.sh calls.
- **[feat]**: K8s: bump renaming tries to 9 for containers. In addition to a new, the renaming script returns labels. There is (close to) 0 sense to collect a cgroup metrics if we don't know its [name/labels](https://github.com/netdata/netdata/blob/3f45709706a59b431b76443b0fbe23c9115768a8/collectors/cgroups.plugin/cgroup-name.sh#L283-L310). We did some tests on a K8s cluster with 100% CPU and it may take 60-70 seconds until K8s api returns data for the container.
- **[feat]**: K8s: cache (store on the host filesystem) /api/v1/pods response and check it before querying K8s API (only for containers). This allows querying the K8s API **much** less frequently.
- **[feat]**: K8s: a cgroup (container) removal doesn't trigger discovery anymore, we run discovery only every `check for new cgroups every`. This behavior is better in K8s.
- **[chore]**: removes `enable cgroup X` option. Not really useful in general. Filtering (enabling/disabling) specific containers is easier using new `enable by default cgroups names matching` feature.
- **[chore]**: a small refactoring of discovery logic + formatting the readme file.

##### Test Plan

- K8s (requires adding debug logging messages)
  - check pause container identification logic.
  - check that cache works (hit/miss).
- run N docker containers, check name resolution works.
- test `enable by default cgroups names matching` filter.

##### Additional Information


<details> <summary>For users: How does this change affect me?</summary>
  
</details>
